token_source.go

```diff
@@ -0,0 +1,133 @@
+package pagerduty
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"log"
+	"os"
+	"strings"
+	"time"
+
+	"golang.org/x/oauth2"
+)
+
+type fileTokenSource struct {
+	base           oauth2.TokenSource
+	configFilePath string
+	clientId       string
+	scopes         []string
+}
+
+type persistedToken struct {
+	*oauth2.Token
+	ClientId string `json:"clientId"`
+	Scopes   string `json:"scopes"`
+}
+
+// NewFileTokenSource creates an oauth2.TokenSource with a Token method which is
+// able to load/save the token info in a file located at configFilePath (e.g.,
+// "token.json" to use the file token.json at CWD).
+func NewFileTokenSource(context context.Context, clientId string, clientSecret string, scopes []string, configFilePath string) oauth2.TokenSource {
+	base := baseTokenSource(context, clientId, clientSecret, scopes)
+
+	fts := &fileTokenSource{
+		base:           base,
+		configFilePath: configFilePath,
+		clientId:       clientId,
+		scopes:         scopes,
+	}
+
+	return oauth2.ReuseTokenSource(nil, fts)
+}
+
+func (c *fileTokenSource) loadToken() (*oauth2.Token, error) {
+	log.Printf("[FTS] Loading token from file\n")
+
+	_, err := os.Stat(c.configFilePath)
+	if os.IsNotExist(err) {
+		return &oauth2.Token{}, nil
+	}
+
+	data, err := os.ReadFile(c.configFilePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to read file %s: %w", c.configFilePath, err)
+	}
+	pt := persistedToken{}
+	if err := json.Unmarshal(data, &pt); err != nil {
+		return nil, fmt.Errorf("failed to decode content of %s: %w", c.configFilePath, err)
+	}
+
+	t := pt.Token
+
+	// [*] TODO: Persisted token is already expired in config file.
```

Comment:
Are these still TODOs? Are we wanting to merge this before those TODOs are complete?